pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -420,33 +434,59 @@ func (c *NPLController) handleAddUpdatePod(key string, obj interface{}) error {
 
 	nplAnnotationsRequired := []NPLAnnotation{}
 
-	// first, check which rules are needed based on the Pod specification (ignoring NPL
-	// annotations) and make sure they are present. As we do so, we build the expected list of
-	// NPL annotations for the Pod.
+	hostPorts := make(map[string]int)
 	for _, container := range podContainers {
 		for _, cport := range container.Ports {
-			port := int(cport.ContainerPort)
-			podPorts[port] = struct{}{}
-			portData := c.portTable.GetEntryByPodIPPort(podIP, int(cport.ContainerPort))
-			if portData == nil { // rule does not exist
-				if int(cport.HostPort) > 0 {
-					klog.V(4).Infof("Host Port is defined for Container %s in Pod %s, thus extra NPL port is not allocated", container.Name, key)
-					nodePort = int(cport.HostPort)
-				} else {
-					nodePort, err = c.portTable.AddRule(podIP, port)
-					if err != nil {
-						return fmt.Errorf("failed to add rule for Pod %s: %v", key, err)
-					}
-				}
+			portProtoStr := cport.Name + delim + string(cport.Protocol)
+			portProtoInt := fmt.Sprint(cport.ContainerPort) + delim + string(cport.Protocol)
+			if int(cport.HostPort) > 0 {
+				klog.V(4).Infof("Host Port is defined for Container %s in Pod %s, thus extra NPL port is not allocated", container.Name, key)
+				hostPorts[portProtoInt] = int(cport.HostPort)
+			}
+			if targetPortsStr.Has(portProtoStr) {
+				targetPortsInt.Insert(portProtoInt)
+			}
+		}
+	}
+
+	if len(targetPortsInt) == 0 {
+		if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+			return err
+		}
+		if _, exists := pod.Annotations[NPLAnnotationKey]; exists {
+			return c.cleanupNPLAnnotationForPod(pod)
+		}
+		return nil
+	}
+
+	// first, check which rules are needed based on the target ports of the Services selecting the Pod
+	// (ignoring NPL annotations) and make sure they are present. As we do so, we build the expected list of
+	// NPL annotations for the Pod.
+	for _, targetPort := range targetPortsInt.List() {
+		portProto := strings.Split(targetPort, delim)
+		port, err := strconv.Atoi(portProto[0])
+		if err != nil {
+			return fmt.Errorf("failed to parse port number from %s for pod %s: %v", portProto[0], key, err)
+		}
```

Comment:
can we have a helper function for building the port:proto string and one for parsing it?